### ISSUE TITLE
Purpose of `baseline` in `EarlyStopping`

### ISSUE THREAD

**Issue by macekmar** [netket/netket#1947]
I am wondering what is the purpose of the parameter `baseline` in [`EarlyStopping`](https://github.com/netket/netket/blob/master/netket/callbacks/earlystopping.py)?

I am writing it here and not in Discussions since I believe there is a bug or at least the docstring is lacking. Actually, I am trying to improve the docstrings of callbacks and do a PR.

With `baseline = None` the callback works as intended: if we do not improve on `loss` in `patience` number of steps, abort the run.

However, the behavior seems strange to me with the parameter `baseline`:

A. We have to drop below `baseline` in `patience` number of steps.
B. Once below, abort if there is no improvement to `loss` in `patience` number of steps.

I think the problem is having the point A with B at the same time (at least with the same `patience`).

Usually, there are three scenarios in learning (example for energy):

1. Good run:
    1. initial slow (or no) convergence, 
    2. fast convergence, 
    4. slow convergence, 
    5. trapped in good local minimum
2. In any point above, too large learning rate and too small preconditioning results in pushing parameters in a wrong minimum having wrong (usually high) energy¹
3. In any point above, too large learning rate and too small preconditioning results in NaN

We wish to stop: too long 1.i, 1.iv, and 2. Case 3 should be stopped with `InvalidLossStopping`. What do we wish from `EarlyStopping`?  Do we wish to stop:

1. case 2 above: I think, it can be stopped without the baseline because in such cases the energy usually does not improve much since we are in a minimum, just a bad one. If we break out of it and begin improving, then the next point follows.
2. too long initial slow convergence (point 1.i above): Say, one demands that we drop in `patience=100` steps below $0.5 E(n_{iter}=0)$ otherwise stop the run because something is wrong (e.g. a particularly bad draw of parameters). This is point A above. Then, all the code about `_best_val` is redundant.
3. In local minimum: Why should we care about the baseline? If the convergence is stopped, just abort the run, right? Yet, we do not want to interrupt the initial slow convergence part. Then the code is wrong because being above the `baseline` also aborts the run. One could use large `patience` for bridging this initial period, but then we are using `patience` for two things.

My suggestions:

1. I think that one could use two stoppers: one without baseline for point 2/B, one with baseline for point 3/A and use and/or between them. We put that into docstrings.
2. Rewrite code so that we only consider point 3/A: only being below `baseline` triggers stopping.
3. Also consider point 2/B. Rather, add another parameter `initial_patience`. AFAIK, the initial slow convergence takes negligible amount of time or result in cases 2 or 3 anyway. Perhaps `initial_patience == patience` is a good setting?

I suggest doing point 2. However, I do not have many experiences with large and long runs. 

Note 1: [The test](https://github.com/netket/netket/blob/master/test/callbacks/test_callbacks.py) is not revealing:
```python
def test_earlystopping_with_baseline():
    baseline = -10
    es = nk.callbacks.EarlyStopping(baseline=baseline)
    vmc = _vmc()

    vmc.run(20, callback=es)
    # We should actually assert something..
```

Note 2: The docstring
```python
            """baseline: Baseline value for the monitored quantity. Training
                will stop if the driver hits the baseline."""
```
suggest the following code on L105:
```python
elif self._is_improvement(loss, self.baseline):
```
i.e., stopping if convergence stops or we hit the baseline. We shouldn't do that, since this is the job of `ConvergenceStopping`.

Note 3:
In all callbacks, we use `driver._loss_name`. I guess the neat way would be to do a `@property` `driver.loss` without the setter since it is set in `__init__`?

¹ Like in this code:
```python
import netket as nk
# Find ground state
N = 6
hilbert = nk.hilbert.Spin(s=1/2, N=N)
graph = nk.graph.Chain(length=N, pbc=True)
ham = nk.operator.Ising(hilbert=hilbert, graph=graph, h=1, J=1)

model = nk.models.RBM(alpha=1)
vs = nk.vqs.FullSumState(hilbert, model)
vs.init(0)
sgd = nk.optimizer.Sgd(0.1)
sr = nk.optimizer.SR(diag_shift=0.001)
gs = nk.driver.VMC(ham, optimizer=sgd, variational_state=vs, preconditioner=sr)
gs.run(n_iter=1000, out='test')
```

**Comment by PhilipVinc** [comment]
cc @alleSini99 and @ lgravina1997  who maybe knows a bit this stuff

**Comment by PhilipVinc** [comment]
We should check what PyTorch/keras do.

**Comment by alleSini99** [comment]
Ok, I see. I would propose an easy fix: remove `baseline` in `EarlyStopping` since already `ConvergenceStopping` stops the optimisation when a value is reached.

**Comment by macekmar** [comment]
It seems `EarlyStopping` is adopted from Keras ([code](https://github.com/keras-team/keras/blob/v3.6.0/keras/src/callbacks/early_stopping.py#L10), [documentation](https://keras.io/api/callbacks/early_stopping/)).
Keras has an additional parameter `start_from_epoch` like in my suggestion 3 above.

I couldn't find callbacks for PyTorch. There is _lightning_ built on Torch which also has [EarlyStopping](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.callbacks.EarlyStopping.html#lightning.pytorch.callbacks.EarlyStopping). It has two thresholds: `stopping_threshold` and `divergence_threshold`. The former acts like NetKet's `ConvergenceStopping` and the latter as the `baseline`.

I guess the best way to do it is to keep `baseline` and add `start_from_iter` with the default value `patience`. In this way, there should be no changes in the behavior of the code used in practice.

**Comment by PhilipVinc** [comment]
Seems reasonable.
@macekmar but do we need to keep EarlyStopping and ConvergenceStopping separated?

**Comment by macekmar** [comment]
I wouldn't join them, since the user can already provide a list of callbacks.
Merger would make `EarlyStopping` even more complicated¹ and testing the run for three different things.

¹ A simple solution would be to change the code into `elif NOT self._is_improvement(loss, self.baseline):` and add a smoothing window. Then `baseline` would act as a threshold. However, this would be a breaking change.

**Comment by PhilipVinc** [comment]
closed by #1949
